lib/ai/embedding.go

```diff
@@ -0,0 +1,165 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ai
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sashabaranov/go-openai"
+
+	embeddingpb "github.com/gravitational/teleport/api/gen/proto/go/teleport/embedding/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	maxOpenAIEmbeddingsPerRequest = 1000
+	EmbeddingPeriod               = 5 * time.Minute
+)
+
+// Sha256Hash is the hash of the embedded content. This hash allows to detect if
+// the embedding is still up-to-date or if the content changed and the resource
+// must be re-embedded.
+type Sha256Hash = [32]byte
+
+// Vector32 is an array of float64 that contains the result of the
+// embedding process. OpenAI client returns []float32, hence Vector32 is the
+// main type for handling vector data.
+type Vector32 = []float32
+
+// Vector64 is an array of float64 that contains the result of the embedding
+// process. While OpenAI returns us 32-bit floats, the vector index uses methods
+// requiring 64-bit floats.
+type Vector64 = []float64
+
+// Embedding contains a Teleport resource embedding. Embeddings are small semantic
+// representations of larger and more complex data. Embeddings can be compared,
+// the smaller the distance between two vectors, the closer the concepts are.
+// Teleport Assist embeds resources to perform semantic search.
+// The Embedding is named after the embedded resource id and kind. For example
+// the SSH node "bastion-01" has the embedding "node/bastion-01".
+type Embedding struct {
+	*embeddingpb.Embedding
+}
+
+// GetEmbeddedKind returns the kind of the resource that was embedded.
+func (e Embedding) GetEmbeddedKind() string {
+	nameParts := strings.Split(e.GetName(), "/")
+	// if name is malformed we bail out
+	if len(nameParts) < 2 {
+		return ""
+	}
+	return nameParts[0]
+}
+
+// GetName returns the Embedding name, composed of the embedded resource kind
+// and the embedded resource ID.
+func (e Embedding) GetName() string {
+	return e.GetMetadata().GetName()
+}
+
+// GetEmbeddedID returns the ID of the resource that was embedded.
+func (e Embedding) GetEmbeddedID() string {
+	nameParts := strings.Split(e.GetName(), "/")
+	// if name is malformed we bail out
+	if len(nameParts) < 2 {
+		return ""
+	}
+	return strings.Join(nameParts[1:], "/")
+}
+
+// NewEmbedding is an Embedding constructor.
+func NewEmbedding(kind, id string, vector Vector64, hash Sha256Hash) *Embedding {
+	return &Embedding{
+		Embedding: &embeddingpb.Embedding{
+			Metadata: &types.Metadata{
+				Name: kind + "/" + id,
+			},
```

Comment:
I was thinking about when I was working with this code. Is there a reason to keep the metadata here instead of just kind and id? Metadata keeps more information such as namespace, labels, etc. That information is never populated, and I find myself getting the empty value multiple times.